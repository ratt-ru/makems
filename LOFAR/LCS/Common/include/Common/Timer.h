//# Timer.h: Accurate timer
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: Timer.h 12871 2009-03-12 12:54:57Z gels $

#ifndef LOFAR_COMMON_TIMER_H
#define LOFAR_COMMON_TIMER_H

// \file
// Very accurate timer for elapsed times.

#include <cstdlib>
#include <cstring>
#include <iostream>

#include <Common/LofarLogger.h>

#if defined __ia64__ && defined __INTEL_COMPILER
#include <ia64regs.h>
#endif


namespace LOFAR {

  // Low-overhead and high-resolution interval timer for use on i386, x86_64,
  // ia64, and powerpc platforms, using the processor's timestamp counter that
  // is incremented each cycle.
  // Put timer.start() and timer.stop() calls around the piece of
  // code to be timed; make sure that start() and stop() calls alternate.
  // A timer can be started and stopped multiple times; both the average and
  // total time, as well as the number of iterations are printed.
  // The measured time is real time (as opposed to user or system time).
  // The timer can be used to measure from 10 nanosecond to a century interval.

  class NSTimer {
  public:
    // Construct.
    // The given name will be used when the timer is printed.
    NSTimer (const std::string& name = std::string(),
	     bool print_on_destruction = false,
	     bool log_on_destruction = false);

    // Destruct.
    // The time is printed on stderr if print_on_destruction is true.
    ~NSTimer();

    // Start the timer.
    void start();
    // Stop the timer
    void stop();

    // Reset the timer to zero.
    void reset();

    // Print the timer.
    // <group>
    std::ostream& print (std::ostream &) const;
    friend std::ostream& operator<< (std::ostream&, const NSTimer&);
    // </group>

    // Get the elapsed time (in seconds).
    double getElapsed() const;

    // Get the total number of times start/stop is done.
    unsigned long long getCount() const;

  private:
    void print_time(std::ostream &, const char *which, double time) const;

    union {
      long long	   total_time;
      struct {
#if defined __PPC__
	int	   total_time_high, total_time_low;
#else
	int	   total_time_low, total_time_high;
#endif
      };
    };

#if defined __i386__ && defined __INTEL_COMPILER && defined _OPENMP
    union {
      unsigned long long count;
      struct {
	int count_low, count_high;
      };
    };
#else
    unsigned long long count;
#endif

    std::string itsName;
    bool print_on_destruction;
    bool log_on_destruction;

    static double CPU_speed_in_MHz;

    static double get_CPU_speed_in_MHz();
  };



  inline void NSTimer::reset()
  {
    total_time = 0;
    count      = 0;
  }

  inline unsigned long long NSTimer::getCount() const
  {
    return count;
  }

  inline NSTimer::NSTimer(const std::string& name, bool print_on_destruction, bool log_on_destruction)
    :
    itsName(name),
    print_on_destruction(print_on_destruction),
    log_on_destruction(log_on_destruction)
  {
    reset();
  }

  inline NSTimer::~NSTimer()
  {
    if (print_on_destruction) {
      if (log_on_destruction) {
        std::stringstream logStr;
        LOG_DEBUG_STR(print(logStr).rdbuf());
      } else
        print(std::clog);
    }
  }

  inline std::ostream& operator<< (std::ostream& str, const NSTimer& timer)
  {
    return timer.print(str);
  }


  inline void NSTimer::start()
  {
#if defined __x86_64__ && defined __INTEL_COMPILER && defined _OPENMP
    asm volatile
    (
	"rdtsc\n\t"
	"shlq $32,%%rdx\n\t"
	"leaq (%%rax,%%rdx),%%rax\n\t"
	"lock;subq %%rax,%0"
    :
	"+m" (total_time)
    :
    :
	"rax", "rdx"
    );
#elif defined __i386__ && defined __INTEL_COMPILER && defined _OPENMP
    asm volatile
    (
	"rdtsc\n\t"
	"lock;subl %%eax,%0\n\t"
	"lock;sbbl %%edx,%1"
    :
	"+m" (total_time_low), "+m" (total_time_high)
    :
    :
	"eax", "edx"
    );
#elif (defined __i386__ || defined __x86_64__) && (defined __GNUC__ || defined __INTEL_COMPILER)
    asm volatile
    (
	"rdtsc\n\t"
	"subl %%eax, %0\n\t"
	"sbbl %%edx, %1"
    :
	"+m" (total_time_low), "+m" (total_time_high)
    :
    :
	"eax", "edx"
    );
#elif (defined __i386__ || defined __x86_64__) && defined __PATHSCALE__
    unsigned eax, edx;

    asm volatile ("rdtsc" : "=a" (eax), "=d" (edx));

    total_time -= ((unsigned long long) edx << 32) + eax;
#elif defined __ia64__ && defined __INTEL_COMPILER
    total_time -= __getReg(_IA64_REG_AR_ITC);
#elif defined __ia64__ && defined __GNUC__
    long long time;
    asm volatile ("mov %0=ar.itc" : "=r" (time));
    total_time -= time;
#elif defined __PPC__ && (defined __GNUC__ || defined __xlC__)
    int high, low, retry;

    asm
    (
	"0:\n\t"
	"mfspr %0,269\n\t"
	"mfspr %1,268\n\t"
	"mfspr %2,269\n\t"
	"cmpw %2,%0\n\t"
	"bne 0b\n\t"
	"subfc %3,%1,%3\n\t"
	"subfe %4,%0,%4"
    :
	"=r" (high), "=r" (low), "=r" (retry),
	"=r" (total_time_low), "=r" (total_time_high)
    :
	"3" (total_time_low), "4" (total_time_high)
    );
#endif
  }


  inline void NSTimer::stop()
  {
#if defined __x86_64__ && defined __INTEL_COMPILER && defined _OPENMP
    asm volatile
    (
	"rdtsc\n\t"
	"shlq $32,%%rdx\n\t"
	"leaq (%%rax,%%rdx),%%rax\n\t"
	"lock;addq %%rax,%0"
    :
	"+m" (total_time)
    :
    :
	"rax", "rdx"
    );
#elif defined __i386__ && defined __INTEL_COMPILER && defined _OPENMP
    asm volatile
    (
	"rdtsc\n\t"
	"lock;addl %%eax, %0\n\t"
	"lock;adcl %%edx, %1"
    :
	"+m" (total_time_low), "+m" (total_time_high)
    :
    :
	"eax", "edx"
    );
#elif (defined __i386__ || defined __x86_64__) && (defined __GNUC__ || defined __INTEL_COMPILER)
    asm volatile
    (
	"rdtsc\n\t"
	"addl %%eax, %0\n\t"
	"adcl %%edx, %1"
    :
	"+m" (total_time_low), "+m" (total_time_high)
    :
    :
	"eax", "edx"
    );
#elif (defined __i386__ || defined __x86_64__) && defined __PATHSCALE__
    unsigned eax, edx;

    asm volatile ("rdtsc\n\t" : "=a" (eax), "=d" (edx));
    total_time += ((unsigned long long) edx << 32) + eax;
#elif defined __ia64__ && defined __INTEL_COMPILER
    total_time += __getReg(_IA64_REG_AR_ITC);
#elif defined __ia64__ && defined __GNUC__
    long long time;
    asm volatile ("mov %0=ar.itc" : "=r" (time));
    total_time += time;
#elif defined __PPC__ && (defined __GNUC__ || defined __xlC__)
    int high, low, retry;

    asm
    (
	"0:\n\t"
	"mfspr %0,269\n\t"
	"mfspr %1,268\n\t"
	"mfspr %2,269\n\t"
	"cmpw %2,%0\n\t"
	"bne 0b\n\t"
	"addc %3,%3,%1\n\t"
	"adde %4,%4,%0"
    :
	"=r" (high), "=r" (low), "=r" (retry),
	"=r" (total_time_low), "=r" (total_time_high)
    :
	"3" (total_time_low), "4" (total_time_high)
    );
#endif

#if defined __x86_64__ && defined __INTEL_COMPILER && defined _OPENMP
    asm volatile ("lock;addq $1,%0" : "+m" (count));
#elif defined __i386__ && defined __INTEL_COMPILER && defined _OPENMP
    asm volatile
    (
	"lock;addl $1,%0\n\t"
	"lock;adcl $0,%1"
    :
	"+m" (count_low), "+m" (count_high)
    );
#else
    ++ count;
#endif
  }

}  // end namespace LOFAR


#endif
