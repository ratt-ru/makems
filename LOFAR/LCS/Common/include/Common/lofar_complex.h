//# lofar_complex.h:
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: lofar_complex.h 11436 2008-07-15 12:16:29Z romein $

#ifndef LOFAR_COMMON_COMPLEX_H
#define LOFAR_COMMON_COMPLEX_H

// \file

//# Put sin, etc. in LOFAR namespace.
#include <Common/lofar_math.h>

//# ICC supports builtin complex for floating point types only.
//# GCC supports it for both floating point and integer types.
//# Both use __complex__ as the complex type specifier.
//# Other compilers might use _Complex, so define LOFAR_BUILTIN_COMPLEX
//# as such to use it.
//# Note that ComplexBuiltin.h uses __real__ and __imag__. Maybe they
//# also need different names for _Complex.
//# Note: include StdInt and BuiltinFP in this order, otherwise the gcc
//# compiler complains that conj already exists in LOFAR namespace.
#if defined __INTEL_COMPILER || defined HAVE_BGL
# define LOFAR_BUILTIN_COMPLEXFP  __complex__
# include <Common/ComplexStdInt.h>
# include <Common/ComplexBuiltinFP.h>

#elif defined __GNUC__ && __GNUC__ < 4 && !defined __INSURE__
# define LOFAR_BUILTIN_COMPLEXFP  __complex__
# define LOFAR_BUILTIN_COMPLEXINT __complex__
# include <Common/ComplexBuiltinFP.h>
# include <Common/ComplexBuiltinInt.h>

#else
# include <Common/ComplexStdFP.h>
# include <Common/ComplexStdInt.h>

#endif

#include <Common/i4complex.h>

namespace LOFAR
{
  // Define complex types in LOFAR namespace.
  using TYPES::i4complex;
  using TYPES::i8complex;
  using TYPES::i16complex;
  using TYPES::u16complex;
  using TYPES::fcomplex;
  using TYPES::dcomplex;

  inline static i4complex makei4complex(i4complex &z) {
    return z;
  }

  inline static i4complex makei4complex(i8complex &z) {
    return makei4complex(real(z) - .5, imag(z) - .5);
  }

  inline static i4complex makei4complex(i16complex &z) {
    return makei4complex(real(z) - .5, imag(z) - .5);
  }

  inline static i4complex makei4complex(u16complex &z) {
    return makei4complex(real(z) - .5, imag(z) - .5);
  }

  inline static i4complex makei4complex(fcomplex &z) {
    return makei4complex(real(z), imag(z));
  }

  inline static i4complex makei4complex(dcomplex &z) {
    return makei4complex(real(z), imag(z));
  }

  inline static i8complex makei8complex(i4complex &z) {
    return makei8complex((int) real(z), (int) imag(z));
  }

  inline static i8complex makei8complex(i8complex &z) {
    return z;
  }

  inline static i8complex makei8complex(i16complex &z) {
    return makei8complex(real(z), imag(z));
  }

  inline static i8complex makei8complex(u16complex &z) {
    return makei8complex(real(z), imag(z));
  }

  inline static i8complex makei8complex(fcomplex &z) {
    return makei8complex((int) real(z), (int) imag(z));
  }

  inline static i8complex makei8complex(dcomplex &z) {
    return makei8complex((int) real(z), (int) imag(z));
  }

  inline static i16complex makei16complex(i4complex &z) {
    return makei16complex((int) real(z), (int) imag(z));
  }

  inline static i16complex makei16complex(i8complex &z) {
    return makei16complex(real(z), imag(z));
  }

  inline static i16complex makei16complex(i16complex &z) {
    return z;
  }

  inline static i16complex makei16complex(u16complex &z) {
    return makei16complex(real(z), imag(z));
  }

  inline static i16complex makei16complex(fcomplex &z) {
    return makei16complex((int) real(z), (int) imag(z));
  }

  inline static i16complex makei16complex(dcomplex &z) {
    return makei16complex((int) real(z), (int) imag(z));
  }

  inline static u16complex makeu16complex(i4complex &z) {
    return makeu16complex((int) real(z), (int) imag(z));
  }

  inline static u16complex makeu16complex(i8complex &z) {
    return makeu16complex(real(z), imag(z));
  }

  inline static u16complex makeu16complex(i16complex &z) {
    return makeu16complex(real(z), imag(z));
  }

  inline static u16complex makeu16complex(u16complex &z) {
    return z;
  }

  inline static u16complex makeu16complex(fcomplex &z) {
    return makeu16complex((unsigned) real(z), (unsigned) imag(z));
  }

  inline static u16complex makeu16complex(dcomplex &z) {
    return makeu16complex((unsigned) real(z), (unsigned) imag(z));
  }

  inline static fcomplex makefcomplex(i4complex &z) {
    return makefcomplex(real(z), imag(z));
  }

  inline static fcomplex makefcomplex(i8complex &z) {
    return makefcomplex((float) real(z), (float) imag(z));
  }

  inline static fcomplex makefcomplex(i16complex &z) {
    return makefcomplex((float) real(z), (float) imag(z));
  }

  inline static fcomplex makefcomplex(u16complex &z) {
    return makefcomplex((float) real(z), (float) imag(z));
  }

  inline static fcomplex makefcomplex(fcomplex &z) {
    return z;
  }

  inline static fcomplex makefcomplex(dcomplex &z) {
    return makefcomplex((float) real(z), (float) imag(z));
  }

  inline static dcomplex makedcomplex(i4complex &z) {
    return makedcomplex(real(z), imag(z));
  }

  inline static dcomplex makedcomplex(i8complex &z) {
    return makedcomplex((double) real(z), (double) imag(z));
  }

  inline static dcomplex makedcomplex(i16complex &z) {
    return makedcomplex((double) real(z), (double) imag(z));
  }

  inline static dcomplex makedcomplex(u16complex &z) {
    return makedcomplex((double) real(z), (double) imag(z));
  }

  inline static dcomplex makedcomplex(fcomplex &z) {
    return makedcomplex((double) real(z), (double) imag(z));
  }

  inline static dcomplex makedcomplex(dcomplex &z) {
    return z;
  }
}


#endif
