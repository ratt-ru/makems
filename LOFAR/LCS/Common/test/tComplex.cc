//# tComplex.cc: Test program for complex performance
//#
//# Copyright (C) 2005
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: tComplex.cc 5570 2005-07-01 16:16:29Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/lofar_complex.h>
#include <complex>
#include <Common/Timer.h>

#if defined __SSE2__
#include <xmmintrin.h>
#include <emmintrin.h>
#endif

using namespace std;
using namespace LOFAR;
using namespace LOFAR::TYPES;


inline complex<double> makedblcomplex(double re, double im)
  { return complex<double>(re,im); }
#define MAKEDBLCOMPLEX(re,im) complex<double>(re,im)


// Define a class holding the builtin complex type.
class mdcomplex
{
public:
  mdcomplex() : itsValue(makedcomplex(0,0)) {}
  mdcomplex(double re, double im) : itsValue(makedcomplex(re,im)) {}
  mdcomplex(dcomplex val) : itsValue(val) {}
  mdcomplex operator* (const mdcomplex that) {return mdcomplex(itsValue*that.itsValue);}
  friend mdcomplex conj (const mdcomplex that) {return mdcomplex(LOFAR::conj(that.itsValue));}
private:
  dcomplex itsValue;
};


// Check if the complex operations give correct results.
void check()
{
  {
    complex<double> c1(1,2), c2(3,4);
    cout << c1 << c2 << endl;
    cout << c1*c2 << c1+c2 << c1-c2 << conj(c1*c2) << c1/c2 << endl;
    cout << sin(c1) << endl;
  }
  {
    dcomplex c1a (makedcomplex(1,2));
    dcomplex c1b (makedcomplex(4,5));
    dcomplex c1c = makedcomplex(6,7);
    dcomplex c1d (makedcomplex(7,8));
    cout << c1a << c1b << c1c << c1d << endl;
    complex<double> cc1(1,2);
    complex<double> cc2(makedblcomplex(3,4));
    dcomplex c1 = *(dcomplex*)(&cc1);
    dcomplex c2 = *(dcomplex*)(&cc2);
    //    dcomplex c2 = 3+4i;
    cout << c1 << c2 << endl;
    dcomplex cc;
    cc = LOFAR::conj(c1*c2);
    cout << c1*c2 << c1+c2 << c1-c2 << cc << c1/c2 << endl;
    cout << sin(c1) << endl;

    u16complex u1a (makeu16complex(10,20));
    cout << u1a << endl;
  }
}

void* allocDCArray (double*& ptr, int nrdc)
{
  // Allocate an array of complex of doubles,
  // Take care it is aligned on 16 byte boundary (for the SSE instructions).
  ptr = new double[2*nrdc+1];
  double* ptr1 = ptr;
  // Make sure data is on 16-byte boundary.
  if ((ptrdiff_t)(ptr1) & 0x000F) ptr1++;
  return ptr1;
}

void* allocDCArray (double*& ptr, int nrdc, complex<double> v)
{
  double* ptr1 = static_cast<double*>(allocDCArray(ptr, nrdc));
  for (int i=0; i<2*nrdc; i+=2) {
    ptr1[i]   = real(v);
    ptr1[i+1] = imag(v);
  }
  return ptr1;
}

void freeDCArray (double* ptr)
{
  delete [] ptr;
}

void checkDCArray (const void* ptr, int nrdc, complex<double> v)
{
  const complex<double>* arr = static_cast<const complex<double>*>(ptr);
  for (int i=0; i<nrdc; ++i) {
    if (arr[i] != v) {
      cout << i << " found=" << arr[i] << " expected=" << v << endl;
    }
  }
}

int main (int argc, const char* argv[])
{
  check();
  int nriter = 1;
  int nrval = 1001;
  if (argc > 1) {
    istringstream istr(argv[1]);
    istr >> nriter;
  }
  if (argc > 2) {
    istringstream istr(argv[2]);
    istr >> nrval;
  }
  cout << "nriter=" << nriter << " nrval=" << nrval << endl;
  cout << ">>>" << endl;

  const complex<double> v1(1,2);
  const complex<double> v2(3,4);
  {
    // Test filling of std::complex using macro.
    complex<double>* arr1 = new complex<double>[nrval];
    {
      NSTimer timer("std::complex fi");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr1[i] = MAKEDBLCOMPLEX(i,i);
	}
      }
      timer.stop();
      timer.print (cout);
    }
    delete[] arr1;
  }
  {
    // Test filling of std::complex using inline function.
    complex<double>* arr1 = new complex<double>[nrval];
    {
      NSTimer timer("std::complex ff");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr1[i] = makedblcomplex (i,i);
	}
      }
      timer.stop();
      timer.print (cout);
    }
    delete[] arr1;
  }
  {
    // Test filling of builtin complex using macro.
    dcomplex* arr1 = new dcomplex[nrval];
    {
      NSTimer timer("dcomplex mfill  ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr1[i] = makedcomplex (i,i);
	}
      }
      timer.stop();
      timer.print (cout);
    }
    delete[] arr1;
  }
  {
    // Test filling of dcomplex (usually builtin) using inline function.
    dcomplex* arr1 = new dcomplex[nrval];
    {
      NSTimer timer("dcomplex ffill  ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr1[i] = makedcomplex (i,i);
	}
      }
      timer.stop();
      timer.print (cout);
    }
    delete[] arr1;
  }
  {
    // Test multiplication of std::complex.
    typedef complex<double> MyComplex;
    double *ptr1, *ptr2, *ptr3;
    MyComplex* arr1 = static_cast<MyComplex*>(allocDCArray(ptr1, nrval, v1));
    MyComplex* arr2 = static_cast<MyComplex*>(allocDCArray(ptr2, nrval, v2));
    MyComplex* arr3 = static_cast<MyComplex*>(allocDCArray(ptr3, nrval));
    {
      NSTimer timer("std::complex * ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * arr2[i];
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*v2);
    }
    {
      NSTimer timer("std::complex *~");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * conj(arr2[i]);
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*conj(v2));
    }
    freeDCArray (ptr1);
    freeDCArray (ptr2);
    freeDCArray (ptr3);
  }
  {
    // Test multiplication of a complex built on top of builtin complex..
    typedef mdcomplex MyComplex;
    double *ptr1, *ptr2, *ptr3;
    MyComplex* arr1 = static_cast<MyComplex*>(allocDCArray(ptr1, nrval, v1));
    MyComplex* arr2 = static_cast<MyComplex*>(allocDCArray(ptr2, nrval, v2));
    MyComplex* arr3 = static_cast<MyComplex*>(allocDCArray(ptr3, nrval));
    {
      NSTimer timer("mdcomplex *    ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * arr2[i];
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*v2);
    }
//     {
//       NSTimer timer("mdcomplex *~   ");
//       timer.start();
//       for (int j=0; j<nriter; ++j) {
// 	for (int i=0; i<nrval; ++i) {
// 	  arr3[i] = arr1[i] * conj(arr2[i]));
// 	}
//       }
//       timer.stop();
//       timer.print (cout);
//     }
    freeDCArray (ptr1);
    freeDCArray (ptr2);
    freeDCArray (ptr3);
  }
#ifdef LOFAR_BUILTIN_COMPLEXFP
  {
    // Test multiplication of builtin complex.
    typedef __complex__ double MyComplex;
    double *ptr1, *ptr2, *ptr3;
    MyComplex* arr1 = static_cast<MyComplex*>(allocDCArray(ptr1, nrval, v1));
    MyComplex* arr2 = static_cast<MyComplex*>(allocDCArray(ptr2, nrval, v2));
    MyComplex* arr3 = static_cast<MyComplex*>(allocDCArray(ptr3, nrval));
    {
      NSTimer timer("_Complex *     ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * arr2[i];
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*v2);
    }
    {
      NSTimer timer("_Complex *~    ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * ~arr2[i];
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*conj(v2));
    }
    freeDCArray (ptr1);
    freeDCArray (ptr2);
    freeDCArray (ptr3);
  }
#endif
  {
    // Test multiplication of dcomplex.
    typedef dcomplex MyComplex;
    double *ptr1, *ptr2, *ptr3;
    MyComplex* arr1 = static_cast<MyComplex*>(allocDCArray(ptr1, nrval, v1));
    MyComplex* arr2 = static_cast<MyComplex*>(allocDCArray(ptr2, nrval, v2));
    MyComplex* arr3 = static_cast<MyComplex*>(allocDCArray(ptr3, nrval));
    {
      NSTimer timer("dcomplex *     ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * arr2[i];
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*v2);
    }
    {
      NSTimer timer("dcomplex *~    ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3[i] = arr1[i] * LOFAR::conj(arr2[i]);
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*conj(v2));
    }
    freeDCArray (ptr1);
    freeDCArray (ptr2);
    freeDCArray (ptr3);
  }
  {
    double* arr1r = new double[nrval];
    double* arr1i = new double[nrval];
    double* arr2r = new double[nrval];
    double* arr2i = new double[nrval];
    double* arr3r = new double[nrval];
    double* arr3i = new double[nrval];
    for (int i=0; i<nrval; ++i) {
      arr1r[i] = real(v1);
      arr1i[i] = imag(v1);
      arr2r[i] = real(v2);
      arr2i[i] = imag(v2);
    }
    {
      NSTimer timer("real,imag *    ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3r[i] = arr1r[i] * arr2r[i] - arr1i[i] * arr2i[i];
	  arr3i[i] = arr1i[i] * arr2r[i] + arr1r[i] * arr2i[i];
	}
      }
      timer.stop();
      timer.print (cout);
      complex<double> v(arr3r[0], arr3i[0]);
      checkDCArray (&v, 1, v1*v2);
    }
    {
      NSTimer timer("real,imag *~   ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	for (int i=0; i<nrval; ++i) {
	  arr3r[i] = arr1r[i] * arr2r[i] + arr1i[i] * arr2i[i];
	  arr3i[i] = arr1i[i] * arr2r[i] - arr1r[i] * arr2i[i];
	}
      }
      timer.stop();
      timer.print (cout);
      complex<double> v(arr3r[0], arr3i[0]);
      checkDCArray (&v, 1, v1*conj(v2));
    }
    delete [] arr1r;
    delete [] arr1i;
    delete [] arr2r;
    delete [] arr2i;
    delete [] arr3r;
    delete [] arr3i;
  }
#if defined __SSE2__
  {
    typedef dcomplex MyComplex;
    double *ptr1, *ptr2, *ptr3;
    MyComplex* arr1 = static_cast<MyComplex*>(allocDCArray(ptr1, nrval, v1));
    MyComplex* arr2 = static_cast<MyComplex*>(allocDCArray(ptr2, nrval, v2));
    MyComplex* arr3 = static_cast<MyComplex*>(allocDCArray(ptr3, nrval));
    {
      NSTimer timer("_Complex * SSE ");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	__m128d *src1 = (__m128d *) arr1;
	__m128d *src2 = (__m128d *) arr2;
	__m128d *dest = (__m128d *) arr3;
	size_t  count = nrval - 1;
	for (size_t i = 0; i < count; i += 2) {
	  __m128d ar = _mm_unpacklo_pd(src1[i], src1[i+1]);
	  __m128d ai = _mm_unpackhi_pd(src1[i], src1[i+1]);
	  __m128d br = _mm_unpacklo_pd(src2[i], src2[i+1]);
	  __m128d bi = _mm_unpackhi_pd(src2[i], src2[i+1]);
	  __m128d cr = _mm_sub_pd(_mm_mul_pd(ar, br), _mm_mul_pd(ai, bi));
	  __m128d ci = _mm_add_pd(_mm_mul_pd(ar, bi), _mm_mul_pd(ai, br));
	  dest[i]    = _mm_unpacklo_pd(cr,ci);
	  dest[i+1]  = _mm_unpackhi_pd(cr,ci);
	}
	if ((count & 1) == 0) {         // uneven array length
	  arr3[count] = arr1[count] * arr2[count];
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*v2);
    }
    {
      NSTimer timer("_Complex *~ SSE");
      timer.start();
      for (int j=0; j<nriter; ++j) {
	__m128d *src1 = (__m128d *) arr1;
	__m128d *src2 = (__m128d *) arr2;
	__m128d *dest = (__m128d *) arr3;
	size_t  count = nrval - 1;
	for (size_t i = 0; i < count; i += 2) {
	  __m128d ar = _mm_unpacklo_pd(src1[i], src1[i+1]);
	  __m128d ai = _mm_unpackhi_pd(src1[i], src1[i+1]);
	  __m128d br = _mm_unpacklo_pd(src2[i], src2[i+1]);
	  __m128d bi = _mm_unpackhi_pd(src2[i], src2[i+1]);
	  __m128d cr = _mm_add_pd(_mm_mul_pd(ar, br), _mm_mul_pd(ai, bi));
	  __m128d ci = _mm_sub_pd(_mm_mul_pd(br, ai), _mm_mul_pd(bi, ar));
	  dest[i]    = _mm_unpacklo_pd(cr,ci);
	  dest[i+1]  = _mm_unpackhi_pd(cr,ci);
	}
	if ((count & 1) == 0) {         // uneven array length
	  arr3[count] = arr1[count] * conj(arr2[count]);
	}
      }
      timer.stop();
      timer.print (cout);
      checkDCArray (arr3, nrval, v1*conj(v2));
    }
    freeDCArray (ptr1);
    freeDCArray (ptr2);
    freeDCArray (ptr3);
  }
#endif

  cout << "<<<" << endl;
  return 0;
}
