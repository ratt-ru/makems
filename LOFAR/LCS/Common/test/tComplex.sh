#!/bin/sh
./runctest.sh tComplex 2>&1 > tComplex.log
