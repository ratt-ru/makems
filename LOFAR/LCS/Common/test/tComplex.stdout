(1,2)(3,4)
(-5,10)(4,6)(-2,-2)(-5,-10)(0.44,0.08)
(3.16578,1.9596)
(1,2)(4,5)(6,7)(7,8)
(1,2)(3,4)
(-5,10)(4,6)(-2,-2)(-5,-10)(0.44,0.08)
(3.16578,1.9596)
(10,20)
nriter=1 nrval=1001
>>>
std::complex fi          : avg =   49 us, total =   49 us, count =         1
std::complex ff          : avg = 76.7 us, total = 76.7 us, count =         1
dcomplex mfill           : avg = 12.5 us, total = 12.5 us, count =         1
dcomplex ffill           : avg = 56.6 us, total = 56.6 us, count =         1
std::complex *           : avg =  144 us, total =  144 us, count =         1
std::complex *~          : avg =  234 us, total =  234 us, count =         1
mdcomplex *              : avg =  101 us, total =  101 us, count =         1
_Complex *               : avg = 13.7 us, total = 13.7 us, count =         1
_Complex *~              : avg = 36.1 us, total = 36.1 us, count =         1
dcomplex *               : avg = 15.3 us, total = 15.3 us, count =         1
dcomplex *~              : avg = 71.6 us, total = 71.6 us, count =         1
real,imag *              : avg =   28 us, total =   28 us, count =         1
real,imag *~             : avg = 26.9 us, total = 26.9 us, count =         1
<<<
